fem/pmgbilinearform.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "../config/config.hpp"
+
+#ifdef MFEM_USE_MPI
+
+#include "pmgbilinearform.hpp"
+#include "transfer.hpp"
+
+namespace mfem
+{
+
+ParMultigridBilinearForm::ParMultigridBilinearForm(
+   ParSpaceHierarchy& spaceHierarchy, ParBilinearForm& bf, Array<int>& ess_bdr,
+   int chebyshevOrder)
+   : MultigridBilinearForm()
+{
+   MFEM_VERIFY(bf.GetAssemblyLevel() == AssemblyLevel::PARTIAL,
+               "Assembly level must be PARTIAL");
+
+   MFEM_VERIFY(bf.GetBBFI()->Size() == 0
+               && bf.GetFBFI()->Size() == 0
+               && bf.GetBFBFI()->Size() == 0,
+               "Only domain integrators are currently supported");
+
+   ParMesh* pmesh = spaceHierarchy.GetFESpaceAtLevel(0).GetParMesh();
+   pmesh_lor = new ParMesh(pmesh, 1, BasisType::GaussLobatto);
+   fec_lor =
+      new H1_FECollection(1, pmesh->Dimension(), BasisType::GaussLobatto);
+   fespace_lor = new ParFiniteElementSpace(pmesh_lor, fec_lor);
+   a_lor = new ParBilinearForm(fespace_lor);
+
+   Array<BilinearFormIntegrator*>& dbfi = *bf.GetDBFI();
+   for (int i = 0; i < dbfi.Size(); ++i)
+   {
+      a_lor->AddDomainIntegrator((*bf.GetDBFI())[i]->Copy());
+   }
+
+   a_lor->UsePrecomputedSparsity();
+   a_lor->Assemble();
+
+   essentialTrueDofs.Append(new Array<int>());
+   spaceHierarchy.GetFESpaceAtLevel(0).GetEssentialTrueDofs(
+      ess_bdr, *essentialTrueDofs.Last());
+
+   HypreParMatrix* hypreCoarseMat = new HypreParMatrix();
+   a_lor->FormSystemMatrix(*essentialTrueDofs.Last(), *hypreCoarseMat);
+
+   HypreBoomerAMG* amg = new HypreBoomerAMG(*hypreCoarseMat);
+   amg->SetPrintLevel(-1);
+   amg->SetMaxIter(1);
```

Comment:
This is the default, so I would remove this line. (We might not even want to add the `SetMaxIter()` method in this PR since this is the only place it's used.)

Reply:
I reverted these changes.